docs/development/design/exit-node.md

```diff
@@ -0,0 +1,145 @@
+x.227.102.183:44728     6/CwH/gzz9jdKnxeVpWFivycFQMBkniLeBNFQq0+f04=
+
+$ wget -qO - icanhazip.com
+x.227.102.183
+
+$ sudo nexctl nexd exit-node disable
+Successfully disabled exit node client on this device
+
+# After subscribing to the exit-node, the client is now originating traffic from the exit node.
+$ wget -qO - icanhazip.com
+x.31.152.254
+```
+
+A diagram of the traffic flow is as follows:
+
+```mermaid
+graph LR
+subgraph Nexodus Mesh with All Devices Peered to one another
+Nex_Device_4["Nexd-1 (100.100.0.1)
+exit-node client"]
+Nex_Device_5["Nexd-2 (100.100.0.2)
+exit-node client"]
+Nex_Device_6["Nexd-3 (100.100.0.3)
+exit-node client"]
+Nexodus_Network_Router_B["Nexd-4 Exit Node 
+wg0 100.100.0.4
+eth(n) 172.16.100.10/24"]
+end
+
+subgraph External Network 
+Internet["Network or Internet
+Receives Traffic from
+the Nexodus Network
+as sourced from eth(n)
+172.16.100.10 via SNAT"]
+end
+
+ Nexodus_Network_Router_B-- "eth(n)" --> Internet
+Nex_Device_4 -- wg0 --> Nexodus_Network_Router_B
+Nex_Device_5 -- wg0 --> Nexodus_Network_Router_B
+Nex_Device_6 -- wg0 --> Nexodus_Network_Router_B
+```
+
+### Client Out of Band Services
+
+- Since all traffic is being tunneled there is a high likelihood that at some point the device will lose connectivity to the exit node and become orphaned. Because of that, we are baking in as many safeguards as possible and options to recover connectivity.
+
+1. Nexodus services will travel out of band via the device's interface containing a default route via Linux policy routing. A new routing table is created with a default route pointing to the same default route in the default namespace. Then using combinations of nftables and rules in the Linux RPDB (routing policy database) traffic destined for the OOB services is mangled and forwarded bypassing the Nexodus driver interface (for example wg0). The initial out of band services are:
+    - HTTPS port 443 to the API server's destination IP(s). If an exit node becomes unavailable, the exit node client device would no longer be able to reach the device. This would orphan the node and require a user to run nexctl to disable the exit node subscription. To allow an administrator or user of many devices to be able to dis-enroll a device from the web UI, we need to maintain api-server connectivity even if an exit node is unavailable.
+    - DNS UDP 53. This should like be discovered via host discovery. There could potentially be scenarios where a DNS is only available via the Nexd mesh but that would be fairly atypical and should be tackled with Nexodus DNS implementation. This rule could also be narrowed if there was a host entry added to the device the agent is running on.
+    - STUN requests, UDP 3478 and 19302 for the RFC defined and Google STUN ports. This is required because once a device is forwarding all traffic through the exit node, the reflexive address for the client device would take on the same reflexive address of the exit node.
```

Comment:
We'll have to document this well. This one seems like it could break other peer-to-peer apps that people intend to use over the top of Nexodus.

Is there a way to mark our own stun requests and only route those differently? That could be tracked as an issue for  later, if needed. I just don't want to forget about it.

Reply:
Here is the basic lab scenario commands for the agent side with annotations I have coded up ready to push:

```
# fwmark set on the interface configuration
REMOTE_KEY=6/CwH/gzz9jdKnxeVpWFivycFQMBkniLeBNFQq0+f04=
sudo wg set wg0 peer $REMOTE_KEY allowed-ips 0.0.0.0/0 persistent-keepalive 25  endpoint 54.227.102.183:41823

# This command enables the src_valid_mark functionality for all network interfaces. This is required for routing marked packets properly with WireGuard.
sudo sysctl -q net.ipv4.conf.all.src_valid_mark=1
# This command adds a rule to the routing policy database (RPDB) that says, "If a packet does not have the firewall mark 51820, look up the routing table 51820."
sudo ip -4 rule add not fwmark 51820 table 51820
# This command adds a rule to the RPDB that says, "When looking up the main routing table, ignore the source address prefix length." This is useful for avoiding unnecessary routing cache updates when using policy-based routing.
sudo ip -4 rule add table main suppress_prefixlength 0
# This command adds a default route to the routing table 51820, which says that all traffic should be sent through the WireGuard interface (wg0).
sudo ip -4 route add 0.0.0.0/0 dev wg0 table 51820

# Nftables

# These commands create an nftables table nexodus-stun-mangle and a chain OUTPUT within it. The purpose of this chain is to mangle (alter) packets. It adds a rule that sets the mark 0x4B66 for packets with UDP destination port 19302.
sudo nft add table ip nexodus-stun-mangle
sudo nft add chain ip nexodus-stun-mangle OUTPUT { type route hook output priority mangle\; policy accept\; }
sudo nft add rule ip nexodus-stun-mangle OUTPUT meta l4proto udp udp dport 19302 counter mark set 0x4B66
# These commands create an nftables table nexodus-stun-snat and a chain POSTROUTING within it. The purpose of this chain is to perform source NAT (SNAT) for outgoing packets. It adds a rule that applies masquerading (SNAT) to packets going out through the enp0s1 interface.
sudo nft add table ip nexodus-stun-snat
sudo nft add chain ip nexodus-stun-snat POSTROUTING { type nat hook postrouting priority srcnat\; policy accept\; }
sudo nft add rule ip nexodus-stun-snat POSTROUTING oifname "enp0s1" counter masquerade
# This command adds a default route to the routing table 19302, which says that all traffic should be sent through the enp0s1 interface with a gateway at 192.168.64.1.
sudo ip -4 route add 0.0.0.0/0 table 19302 via 192.168.64.1 dev enp0s1
# This command adds a rule to the RPDB that says, "If a packet has the firewall mark 19302, look up the routing table 19302." This is used to route marked packets with destination port 19302 using the custom routing table created in step 11.
sudo ip -4 rule add fwmark 19302 table 19302

```

Reply:
Resolving as its being discussed later in the review.